toolkit/components/filewatcher/NativeFileWatcherLnx.cpp

```diff
@@ -0,0 +1,1388 @@
+  nsString mResourcePath;
+};
+
+/**
+ * An event used to notify the main thread of a change in a watched
+ * resource.
+ */
+class WatchedChangeEvent final : public Runnable
+{
+public:
+  /**
+   * @param aOnChange The passed change callback.
+   * @param aChangedResource The name of the changed resource.
+   */
+  WatchedChangeEvent(
+    const nsMainThreadPtrHandle<nsINativeFileWatcherCallback>& aOnChange,
+    const nsAString& aChangedResource)
+    : Runnable("WatchedChangeEvent")
+    , mOnChange(aOnChange)
+    , mChangedResource(aChangedResource)
+  {
+    MOZ_ASSERT(!NS_IsMainThread());
+  }
+
+  NS_IMETHOD Run() override
+  {
+    MOZ_ASSERT(NS_IsMainThread());
+
+    // The second parameter is reserved for future uses: we use 0 as a
+    // placeholder.
+    (void)mOnChange->Changed(mChangedResource, 0);
+    return NS_OK;
+  }
+
+private:
+  nsMainThreadPtrHandle<nsINativeFileWatcherCallback> mOnChange;
+  nsString mChangedResource;
+};
+
+// Define these callback array types to make the code easier to read.
+typedef nsTArray<nsMainThreadPtrHandle<nsINativeFileWatcherCallback>>
+  ChangeCallbackArray;
+typedef nsTArray<nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback>>
+  ErrorCallbackArray;
+
+static mozilla::LazyLogModule gNativeWatcherPRLog("NativeFileWatcherService");
+#define FILEWATCHERLOG(...)                                                    \
+  MOZ_LOG(gNativeWatcherPRLog, mozilla::LogLevel::Debug, (__VA_ARGS__))
+
+struct WatchedResourceDescriptor
+{
+  // The path on the file system of the watched resource.
+  nsString mPath;
+
+  // The watched descriptor returned from inotify_add_watch().
+  int mWatchedResourceDescriptor;
+
+  WatchedResourceDescriptor(const nsAString& aPath, const int anHandle)
+    : mPath(aPath)
+    , mWatchedResourceDescriptor(anHandle)
+  {
+  }
+};
+
+/**
+ *  * A structure used to pass the callbacks to the AddPathRunnableMethod() and
+ *   * RemovePathRunnableMethod().
+ *    */
+struct PathRunnablesParametersWrapper
+{
+  nsString mPath;
+  nsMainThreadPtrHandle<nsINativeFileWatcherCallback> mChangeCallbackHandle;
+  nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback> mErrorCallbackHandle;
+  nsMainThreadPtrHandle<nsINativeFileWatcherSuccessCallback>
+    mSuccessCallbackHandle;
+
+  PathRunnablesParametersWrapper(
+    const nsAString& aPath,
+    const nsMainThreadPtrHandle<nsINativeFileWatcherCallback>& aOnChange,
+    const nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback>& aOnError,
+    const nsMainThreadPtrHandle<nsINativeFileWatcherSuccessCallback>&
+      aOnSuccess)
+    : mPath(aPath)
+    , mChangeCallbackHandle(aOnChange)
+    , mErrorCallbackHandle(aOnError)
+    , mSuccessCallbackHandle(aOnSuccess)
+  {
+  }
+};
+
+/**
+ * This runnable is dispatched from the main thread to get the notifications of
+ * the changes in the watched resources by continuously calling the blocking
+ * function GetQueuedCompletionStatus. This function queries the status of the
```

Comment:
I'm pretty sure that this is not what happens on Linux.

Could you make sure that the documentation describes what you're doing?